File: Linear Control Course/Chapter29/Lesson4/Chapter29_Lesson4.java
public class CruiseControlSimulation {

    static class Vehicle {
        double m;   // mass
        double b;   // drag coefficient
        double Ku;  // engine gain
        double v;   // speed state

        Vehicle(double m, double b, double Ku) {
            this.m = m;
            this.b = b;
            this.Ku = Ku;
            this.v = 0.0;
        }

        // One Euler step: dv/dt = (Ku*u - b*v)/m
        void step(double u, double h) {
            double dv = (Ku * u - b * v) / m;
            v += dv * h;
        }
    }

    static class PIController {
        double Kp;
        double Ki;
        double ei;  // integral of error

        PIController(double Kp, double Ki) {
            this.Kp = Kp;
            this.Ki = Ki;
            this.ei = 0.0;
        }

        double compute(double r, double y, double h) {
            double e = r - y;
            ei += e * h;
            return Kp * e + Ki * ei;
        }
    }

    public static void main(String[] args) {
        double m  = 1000.0;
        double b  = 50.0;
        double Ku = 500.0;

        // Example gains
        double Kp = 0.8;
        double Ki = 0.4;

        Vehicle veh = new Vehicle(m, b, Ku);
        PIController ctrl = new PIController(Kp, Ki);

        double h = 0.01;
        double T = 20.0;
        int N = (int) (T / h);

        double v_ref = 10.0; // desired speed [m/s]

        for (int k = 0; k <= N; ++k) {
            double t = k * h;
            double u = ctrl.compute(v_ref, veh.v, h);
            veh.step(u, h);

            if (k % 200 == 0) {
                System.out.println("t = " + t + " s, v = " + veh.v + " m/s");
            }
        }
    }
}

File: Linear Control Course/Chapter29/Lesson4/Chapter29_Lesson4.py
import numpy as np
import matplotlib.pyplot as plt
import control as ct  # pip install control

# Physical parameters (example)
m = 1000.0   # kg
b = 50.0     # N*s/m (effective drag)
Ku = 500.0   # N per unit control input (normalized throttle)

# Desired specs
Mp = 0.1       # max overshoot <= 10%
Ts = 5.0       # 2% settling time around 5 s

# Compute zeta from Mp (approximate inverse of Mp-zeta relation)
# Here we choose a typical damping ratio for 10% overshoot:
zeta = 0.6

# Natural frequency from settling time formula Ts ≈ 4/(zeta*wn)
wn = 4.0 / (zeta * Ts)

# PI gains from section 4 formulas
Kp = (2.0 * zeta * wn * m - b) / Ku
Ki = (wn**2 * m) / Ku

print("PI gains:")
print("Kp =", Kp)
print("Ki =", Ki)

# Define plant G(s) = Ku / (m s + b)
numG = [Ku]
denG = [m, b]
G = ct.TransferFunction(numG, denG)

# Controller C(s) = Kp + Ki/s
C = ct.TransferFunction([Kp, Ki], [1, 0])

# Closed-loop transfer function from reference to speed
T = ct.feedback(C * G, 1)

# Step response for a step in reference speed (1 m/s step)
t = np.linspace(0, 20, 1000)
t, y = ct.step_response(T, t)

plt.figure()
plt.plot(t, y)
plt.axhline(1.0, linestyle="--")
plt.xlabel("Time [s]")
plt.ylabel("Speed [m/s]")
plt.title("Cruise Control: Closed-Loop Step Response (PI)")
plt.grid(True)
plt.show()
